Race condition in LabelIdentityReconciler

**Describe the bug**
<!--
A clear and concise description of what the bug is.
If you believe this bug is a security issue, please don't use this template and follow our [security guidelines](/SECURITY.md)
-->
Unit test failed consistently as the following. And the race condition exists in real world too.
```
I1013 08:23:35.869746   10110 gateway_controller.go:196] "Created a ClusterInfo kind of ResourceExport" clusterinfo="default/cluster-a-clusterinfo"
==================
WARNING: DATA RACE
Read at 0x00c0002fbef0 by goroutine 185:
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).createLabelIdentityResExport()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:265 +0x84
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).syncLabelResourceExport()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:252 +0xe6
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).processLabelForResourceExport()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:225 +0x194
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).labelQueueWorker()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:210 +0x52
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).labelQueueWorker-fm()
      <autogenerated>:1 +0x39
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /root/go/pkg/mod/k8s.io/apimachinery@v0.24.0/pkg/util/wait/wait.go:155 +0x48
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /root/go/pkg/mod/k8s.io/apimachinery@v0.24.0/pkg/util/wait/wait.go:156 +0xce
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /root/go/pkg/mod/k8s.io/apimachinery@v0.24.0/pkg/util/wait/wait.go:133 +0x10d
  k8s.io/apimachinery/pkg/util/wait.Until()
      /root/go/pkg/mod/k8s.io/apimachinery@v0.24.0/pkg/util/wait/wait.go:90 +0x48
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).Run.func2()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:204 +0x4c

Previous write at 0x00c0002fbef0 by goroutine 182:
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).Reconcile()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:92 +0x246
  antrea.io/antrea/multicluster/controllers/multicluster.TestLabelIdentityReconciler.func1()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller_test.go:186 +0x16b9
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Goroutine 185 (running) created at:
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).Run()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:204 +0xef
  antrea.io/antrea/multicluster/controllers/multicluster.TestLabelIdentityReconciler.func1.2()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller_test.go:177 +0x47

Goroutine 182 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x75d
  antrea.io/antrea/multicluster/controllers/multicluster.TestLabelIdentityReconciler()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller_test.go:167 +0x1728
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c0002fbea8 by goroutine 185:
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).getLabelIdentityResourceExport()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:292 +0x84
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).createLabelIdentityResExport()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:266 +0xcc
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).syncLabelResourceExport()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:252 +0xe6
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).processLabelForResourceExport()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:225 +0x194
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).labelQueueWorker()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:210 +0x52
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).labelQueueWorker-fm()
      <autogenerated>:1 +0x39
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /root/go/pkg/mod/k8s.io/apimachinery@v0.24.0/pkg/util/wait/wait.go:155 +0x48
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /root/go/pkg/mod/k8s.io/apimachinery@v0.24.0/pkg/util/wait/wait.go:156 +0xce
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /root/go/pkg/mod/k8s.io/apimachinery@v0.24.0/pkg/util/wait/wait.go:133 +0x10d
  k8s.io/apimachinery/pkg/util/wait.Until()
      /root/go/pkg/mod/k8s.io/apimachinery@v0.24.0/pkg/util/wait/wait.go:90 +0x48
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).Run.func2()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:204 +0x4c

Previous write at 0x00c0002fbea8 by goroutine 182:
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).Reconcile()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:96 +0x2bc
  antrea.io/antrea/multicluster/controllers/multicluster.TestLabelIdentityReconciler.func1()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller_test.go:186 +0x16b9
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Goroutine 185 (running) created at:
  antrea.io/antrea/multicluster/controllers/multicluster.(*LabelIdentityReconciler).Run()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller.go:204 +0xef
  antrea.io/antrea/multicluster/controllers/multicluster.TestLabelIdentityReconciler.func1.2()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller_test.go:177 +0x47

Goroutine 182 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x75d
  antrea.io/antrea/multicluster/controllers/multicluster.TestLabelIdentityReconciler()
      /root/antrea/multicluster/controllers/multicluster/label_identity_controller_test.go:167 +0x1728
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47
==================
--- FAIL: TestLabelIdentityReconciler (0.15s)
    --- FAIL: TestLabelIdentityReconciler/pod_delete_event_no_stale_label (0.03s)
        testing.go:1319: race detected during execution of test
    testing.go:1319: race detected during execution of test
I1013 08:23:36.444407   10110 leader_clusterset_controller.go:81] "Received ClusterSet add/update" clusterset="mcs1/clusterset1"
I1013 08:23:36.444591   10110 controller_utils.go:41] "Validating ClusterClaim" namespace="mcs1"
I1013 08:23:36.446566   10110 controller_utils.go:53] "Processing ClusterClaim" name="clusterset.k8s.io" value="clusterset1"
I1013 08:23:36.446703   10110 controller_utils.go:53] "Processing ClusterClaim" name="id.k8s.io" value="leader1"
I1013 08:23:36.447851   10110 leader_clusterset_controller.go:81] "Received ClusterSet add/update" clusterset="mcs1/clusterset1"
I1013 08:23:36.447897   10110 controller_utils.go:41] "Validating ClusterClaim" namespace="mcs1"
I1013 08:23:36.448184   10110 controller_utils.go:53] "Processing ClusterClaim" name="clusterset.k8s.io" value="clusterset1"
I1013 08:23:36.448227   10110 controller_utils.go:53] "Processing ClusterClaim" name="id.k8s.io" value="leader1"
I1013 08:23:36.448819   10110 leader_clusterset_controller.go:81] "Received ClusterSet add/update" clusterset="mcs1/clusterset1"
I1013 08:23:36.452075   10110 leader_clusterset_controller.go:81] "Received ClusterSet add/update" clusterset="mcs1/clusterset1"
I1013 08:23:36.452200   10110 controller_utils.go:41] "Validating ClusterClaim" namespace="mcs1"
I1013 08:23:36.453132   10110 controller_utils.go:53] "Processing ClusterClaim" name="clusterset.k8s.io" value="clusterset1"
I1013 08:23:36.453266   10110 controller_utils.go:53] "Processing ClusterClaim" name="id.k8s.io" value="leader1"
I1013 08:23:36.454444   10110 leader_clusterset_controller.go:73] "Received ClusterSet delete" clusterset="mcs1/clusterset1"
I1013 08:23:36.457855   10110 leader_clusterset_controller.go:81] "Received ClusterSet add/update" clusterset="mcs1/clusterset1"
I1013 08:23:36.457973   10110 controller_utils.go:41] "Validating ClusterClaim" namespace="mcs1"
I1013 08:23:36.458943   10110 controller_utils.go:53] "Processing ClusterClaim" name="clusterset.k8s.io" value="clusterset1"
I1013 08:23:36.459092   10110 controller_utils.go:53] "Processing ClusterClaim" name="id.k8s.io" value="leader1"
I1013 08:23:36.462563   10110 member_clusterset_controller.go:93] "Received ClusterSet delete" clusterset="default/clusterset1"
I1013 08:23:36.464981   10110 memberclusterannounce_controller.go:281] "Removing member cluster from the ClusterSet" cluster="east" ClusterSet="mcs1/clusterset1"
I1013 08:23:36.465970   10110 memberclusterannounce_controller.go:277] "Member cluster not found in ClusterSet" ClusterSet="clusterset1" cluster="not-exist"
I1013 08:23:36.466704   10110 memberclusterannounce_controller.go:219] "Added member cluster" cluster=east
I1013 08:23:36.467423   10110 memberclusterannounce_controller.go:219] "Added member cluster" cluster=east
I1013 08:23:36.467473   10110 memberclusterannounce_controller.go:227] "Removed member cluster" cluster=east
I1013 08:23:36.468143   10110 memberclusterannounce_controller.go:219] "Added member cluster" cluster=east
I1013 08:23:36.468888   10110 memberclusterannounce_controller.go:219] "Added member cluster" cluster=east
I1013 08:23:36.469307   10110 memberclusterannounce_controller.go:111] "Adding member cluster to ClusterSet" cluster="east"
I1013 08:23:36.469595   10110 memberclusterannounce_controller.go:248] "Member cluster already exists in ClusterSet" cluster="east" ClusterSet="clusterset1"
I1013 08:23:36.469641   10110 memberclusterannounce_controller.go:116] "Adding finalizer to MemberClusterAnnounce" MemberClusterAnnounce="mcs1/member-announce-from-east"
E1013 08:23:38.485828   10110 node_controller.go:121] "There is no valid Gateway IP for Node" err="no valid IP address for Gateway Node node-1" node="node-1"
I1013 08:23:38.488552   10110 node_controller.go:253] "Found good Gateway candidate" node="node-2"
I1013 08:23:38.491429   10110 node_controller.go:253] "Found good Gateway candidate" node="node-2"
E1013 08:23:38.493311   10110 node_controller.go:121] "There is no valid Gateway IP for Node" err="the Gateway IP annotation invalid-gatewayip on Node node-4 is not a valid IP address" node="node-4"
I1013 08:23:38.493727   10110 node_controller.go:253] "Found good Gateway candidate" node="node-2"
I1013 08:23:38.498698   10110 resourceexport_controller.go:232] "Cleaning up ResourceImport" resourceimport="default/default-nginx-service"
I1013 08:23:38.510436   10110 clusterinfoexport_handler.go:81] "Updating ResourceImport" resourceimport="default/cluster-b-default-clusterinfo"
I1013 08:23:38.515647   10110 serviceexport_controller.go:207] "It's not allowed to export the multi-cluster controller auto-generated Service" service="default/antrea-mc-nginx"
I1013 08:23:38.519600   10110 serviceexport_controller.go:300] "Service has new changes, update ResourceExport" service="default/nginx1" resourceexport="default/cluster-a-default-nginx1-service"
I1013 08:23:38.519692   10110 serviceexport_controller.go:569] "Creating ResourceExport" resourceexport="default/cluster-a-default-nginx1-service"
I1013 08:23:38.520042   10110 serviceexport_controller.go:315] "Endpoints have new change, update ResourceExport" endpoints="default/nginx1" resourceexport="default/cluster-a-default-nginx1-endpoints"
I1013 08:23:38.520122   10110 serviceexport_controller.go:569] "Creating ResourceExport" resourceexport="default/cluster-a-default-nginx1-endpoints"
I1013 08:23:38.523271   10110 serviceexport_controller.go:300] "Service has new changes, update ResourceExport" service="default/nginx" resourceexport="default/cluster-a-default-nginx-service"
I1013 08:23:38.523380   10110 serviceexport_controller.go:569] "Creating ResourceExport" resourceexport="default/cluster-a-default-nginx-service"
I1013 08:23:38.523628   10110 serviceexport_controller.go:315] "Endpoints have new change, update ResourceExport" endpoints="default/nginx" resourceexport="default/cluster-a-default-nginx-endpoints"
I1013 08:23:38.523712   10110 serviceexport_controller.go:569] "Creating ResourceExport" resourceexport="default/cluster-a-default-nginx-endpoints"
I1013 08:23:38.525637   10110 serviceexport_controller.go:300] "Service has new changes, update ResourceExport" service="default/nginx" resourceexport="default/cluster-a-default-nginx-service"
I1013 08:23:38.526303   10110 serviceexport_controller.go:315] "Endpoints have new change, update ResourceExport" endpoints="default/nginx" resourceexport="default/cluster-a-default-nginx-endpoints"
I1013 08:23:38.528296   10110 serviceexport_controller.go:300] "Service has new changes, update ResourceExport" service="default/nginx" resourceexport="default/cluster-a-default-nginx-service"
I1013 08:23:38.528977   10110 serviceexport_controller.go:315] "Endpoints have new change, update ResourceExport" endpoints="default/nginx" resourceexport="default/cluster-a-default-nginx-endpoints"
I1013 08:23:38.531056   10110 serviceexport_controller.go:256] "Service has been converted into ResourceExport and no change, skip it" service="default/nginx" resourceexport="default/cluster-a-default-nginx-service"
I1013 08:23:38.531136   10110 serviceexport_controller.go:269] "Endpoints has been converted into ResourceExport and no change, skip it" endpoints="default/nginx" resourceexport="default/cluster-a-default-nginx-endpoints"
I1013 08:23:38.533878   10110 stale_controller.go:184] "Cleaning up stale imported Service" service="default/nginx"
I1013 08:23:38.533965   10110 stale_controller.go:191] "Cleaning up stale ServiceImport" serviceimport="default/nginx"
I1013 08:23:38.537484   10110 stale_controller.go:218] "Cleaning up stale imported ACNP" acnp="antrea-mc-some-deleted-resimp"
I1013 08:23:38.540169   10110 stale_controller.go:301] "Cleaning up stale ResourceExport" ResourceExport="default/cluster-a-default-nginx-endpoint"
I1013 08:23:38.540248   10110 stale_controller.go:301] "Cleaning up stale ResourceExport" ResourceExport="default/cluster-a-default-nginx-service"
I1013 08:23:38.540430   10110 stale_controller.go:368] "Cleaning up stale ClusterInfo kind of ResourceExport" resourceexport="default/cluster-a-clusterinfo"
I1013 08:23:38.541181   10110 stale_controller.go:344] "Cleaning up stale ResourceExport" ResourceExport="default/cluster-a-fe2634b306c51bd3"
I1013 08:23:38.542801   10110 stale_controller.go:243] "Cleaning up stale ClusterInfoImport" clusterinfoimport="default/cluster-b-default-clusterinfo"
I1013 08:23:38.545061   10110 stale_controller.go:389] "Cleaning up stale MemberClusterAnnounce. It has not been updated within the agreed period" MemberClusterAnnounce="default/member-cluster-from-cluster-outdated" agreedPeriod="5m0s"
```

--- a/multicluster/controllers/multicluster/label_identity_controller.go
+++ b/multicluster/controllers/multicluster/label_identity_controller.go
@@ -52,6 +52,7 @@ type (
 	LabelIdentityReconciler struct {
 		client.Client
 		Scheme           *runtime.Scheme
+		commonAreaMutex  sync.Mutex
 		commonAreaGetter RemoteCommonAreaGetter
 		remoteCommonArea commonarea.RemoteCommonArea
 		// labelMutex prevents concurrent access to labelToPodsCache and podLabelCache.
@@ -87,16 +88,11 @@ func NewLabelIdentityReconciler(
 // +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
 func (r *LabelIdentityReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	klog.V(2).InfoS("Reconciling Pod for label identity", "pod", req.NamespacedName)
-	var commonArea commonarea.RemoteCommonArea
-	var err error
-	commonArea, r.localClusterID, err = r.commonAreaGetter.GetRemoteCommonAreaAndLocalID()
-	if commonArea == nil {
-		return ctrl.Result{Requeue: true}, err
+	if requeue := r.checkRemoteCommonArea(); requeue {
+		return ctrl.Result{Requeue: true}, nil
 	}
-	r.remoteCommonArea = commonArea
 	var pod v1.Pod
 	var ns v1.Namespace
-
 	if err := r.Client.Get(ctx, req.NamespacedName, &pod); err != nil {
 		if apierrors.IsNotFound(err) {
 			klog.V(2).InfoS("Pod is deleted", "pod", req.NamespacedName)
@@ -112,6 +108,22 @@ func (r *LabelIdentityReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 	normalizedLabel := getNormalizedLabel(ns.Labels, pod.Labels, ns.Name)
 	r.onPodCreateOrUpdate(req.NamespacedName.String(), normalizedLabel)
 	return ctrl.Result{}, nil
+}
+
+// checkRemoteCommonArea initializes remoteCommonArea for the reconciler if necessary,
+// or tells the Reconcile function to requeue if the remoteCommonArea is not ready.
+func (r *LabelIdentityReconciler) checkRemoteCommonArea() bool {
+	r.commonAreaMutex.Lock()
+	defer r.commonAreaMutex.Unlock()
+
+	if r.remoteCommonArea == nil {
+		commonArea, localClusterID, _ := r.commonAreaGetter.GetRemoteCommonAreaAndLocalID()
+		if commonArea == nil {
+			return true
+		}
+		r.remoteCommonArea, r.localClusterID = commonArea, localClusterID
+	}
+	return false
 }
 
 // SetupWithManager sets up the controller with the Manager.

